### ISSUE TITLE
docs not loading properly in Safari

### ISSUE THREAD

**Issue by frjnn** [rust-lang/rust#85444]
## Setup

Command issued: `cargo doc --open`
Safari version: 14.1
MacOS: BigSur 11.3.1

rustup version: 1.24.2
rustc version: 1.5.2

I expected to see the crate's documentation being displayed and formatted properly.

Instead, this happened:
<img width="1417" alt="Screenshot 2021-05-18 at 19 58 29" src="https://user-images.githubusercontent.com/54247008/118701386-f2a94100-b813-11eb-8874-1a1e0b039643.png">

This is **not** by any mean a problem of ndarray, I'm experiencing this with all the crates I've tried so far.

**Comment by klensy** [comment]
>rustc version: 1.5.2

1.52? Try 1.52.1. Press F12 in browser and check any errors (is safari have any dev tools like chrome?)

**Comment by frjnn** [comment]
My bad! rustc's version is 1.52.1, I also corrected the issue.

**Comment by frjnn** [comment]
@estebank @klensy enabling the develop menu seems to have solved the issue. Now the documentation loads properly.

What I did:
1. Opened Safari
2. Went to Preferences
3. Went to Advanced Preferences
4. Check "Show Develop menu in menu bar"

I don't know if this can be closed or not though.

**Comment by klensy** [comment]
I don't think that docs should require enabling devtools in browser, perhaps you rebooted browsed or OS? I don't have safari, so can't check that issue.

**Comment by frjnn** [comment]
No, I didn't reboot anything, I didn't ever restart Safari, I just enabled the Develop Menu.

**Comment by AlvinKuruvilla** [comment]
Second, this issue op, been having it for a while now. Wondering if there is a fix or a way to change what browser cargo-doc uses?

**Comment by apiraino** [comment]
@AlvinKuruvilla I think rustdoc opens the system default browser. Also, is this issue specific to Safari? Do you have any other browser installed that you can temporarily set as default to test this?

**Comment by AlvinKuruvilla** [comment]
Yes I changed my default browser to brave and now the docs appear correctly

**Comment by DarrienG** [comment]
Confirming this is still an issue in the latest version of rust (1.55.0). Enabling the develop menu fixes it.

**Comment by emabee** [comment]
Same here - it works in firefox, but not in Safari, unless I enable the dev tools; rust version (for doc) is rustc 1.60.0-nightly (75d9a0ae2 2022-02-16)

**Comment by Enselic** [comment]
This is caused by Safari disallowing by default to load external files from disk such as `../rustdoc.css`. When Develop Menu is enabled, this restriction is lifted. Manually inlining the CSS into `index.html` is another way to make the styling work.

**Comment by jsha** [comment]
Thanks for the details, @Enselic! Given that, I don't think we can work around it in rustdoc. We rely heavily on external files, and inlining everything would be a performance problem for viewing the docs on the web. It would also be a size problem for locally-generated docs.

Perhaps we could have `cargo doc --open` and `rustup doc` detect if the system browser is Safari, and if so try Firefox instead?

We could also include inline script at the top of each page that checks for `file:///` URL and Safari browser, and injects a banner explaining why the page isn't rendering right and what to do about it.

**Comment by ptdecker** [comment]
I have developer menus enabled but am experiencing the same issue.  There is also someone on StackOverflow who has put it up as a question:  [link](https://stackoverflow.com/questions/73559476/cargo-doc-run-does-not-load-the-doc-in-safari-properly).  Attached is a screenshot.  

![safarierror](https://user-images.githubusercontent.com/537053/188294530-1f75fcaf-f522-4fc7-bffb-f44aabf24d69.jpg)

**Comment by jsha** [comment]
@ptdecker try this:

```
python -mhttp.server --directory target/doc/ 5555 &
open http://localhost:5555/guessing_game/
```

This starts a web server locally that serves the generated documentation, then opens it in your default browser (presumably Safari). If that works, it might be another way we can work around the issue for Safari users.

**Comment by mredigonda** [comment]
This is not just a Safari issue, it also doesn't work on Chrome in my Mac M1.

```
(...) from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, chrome-untrusted, https, isolated-app.
```

These are the kinds of errors I get.

Basically it seems that the protocol `file://` is not allowed to load resources like fonts, css, etc.

The workaround to open a server with Python works.

**Comment by nixpulvis** [comment]
:(

**Comment by kovar** [comment]
On macOS Sonoma (14.1.1), Safari 17.1, you need to go to Settings and then the Advanced tab and enable

- [x] Show features for web developers

A new Developer tab becomes available, where you check

- [x] Disable local file restrictions

See image below 👇 

<img width="948" alt="image" src="https://github.com/rust-lang/rust/assets/72451476/64409c8e-b1cb-47f7-95a3-ebad3c20d184">

**Comment by estebank** [comment]
I believe @jsha's solution of starting a lightweight server might be the best way to address this issue going forward, but we'd need to test that we don't add other problems (because of OS-specific security features that would cause the server not to start or give scary warnings).

**Comment by shanecelis** [comment]
Setting up new computer on macOS 14.3. Turning on developer mode by itself did not fix my issue. But turning on developer mode then quitting Safari did fix it. I did not have to disable local file restrictions.

**Comment by RodBurman** [comment]
With macOS Sequoia 15.0 the problem persists and can still be fixed in the developer options of safari. If injecting the css into index.html fixes the problem (as suggested in the comments above) then maybe do that? If this bloats things too much for normal usage "we" could addd another cargo doc option e.g. --css-in-index?

**Comment by nixpulvis** [comment]
Personally, enabling the developer mode for safari is a better solution for me than starting a webserver. That said, discoverability is an issue. I wouldn't hate a small browser detection script which showed a note about enabling this mode on safari for users with it disabled and causing issues.

This really is pretty bad default user experience.

**Sidenote:**

Also, am I the only one wondering why this "security feature" is so easily bypassed by a global Enable Developer Mode setting? Don't so many people run with this setting. If this was a real security issue, aren't all these people thus vulnerable? Seems wrong, but I'm honestly not even sure what the attack vector here is exactly. Also, why don't other browsers have this issue?

**Comment by RodBurman** [comment]
I cannot replicate the problem with the now current version of cargo 1.88.0 running on (now current) macOS Sequoia 15.5. So to me the problem seems to have been fixed since my last post in October. This is with developer tools switched off in Safari.
